File: src/Factoriod.Models/Game/ServerSettings.cs
using System.ComponentModel.DataAnnotations;
using Factoriod.Utilities;

namespace Factoriod.Models.Game;

/// <summary>
/// The server-settings.json file is the primary settings file for the server.
/// </summary>
/// <param name="Name">Name of the game as it will appear in the game listing.</param>
/// <param name="Description">Description of the game that will appear in the listing.</param>
/// <param name="Tags">Tags of the game that will appear in the listing.</param>
/// <param name="MaxPlayers">Maximum number of players allkowed, admins can join even a full server. Defaults to unlimited.</param>
/// <param name="Visibility">Visibility of the game.</param>
/// <param name="Username">Your factorio.com login credentials. Required for games with <see cref="Visibility.Public"/></param>
/// <param name="RequireUserVerification">When set to true, the server will only allow clients that have a valid Factorio.com account</param>
/// <param name="MaxUploadInKilobytesPerSecond">Maximum number of KB/s that the server will upload. Defaults to unlimited.</param>
/// <param name="MaxUploadSlots">Maximum number of slots the server will upload./param>
/// <param name="MinimumLatencyInTicks">Smallest amount of latency the game can provide. One tick is 16ms.</param>
/// <param name="MaxHeartbeatsPerSecond">Network tick rate. Maximum rate game updates packets are sent at before bundling them together.</param>
/// <param name="IgnorePlayerLimitForReturningPlayers">Players that played on this map already can join even when the max player limit was reached.</param>
/// <param name="AllowCommands">Whether commands are allowed. Possible value are true, false, and admins-only.</param>
/// <param name="AutosaveInterval">Autosave interval in minutes.</param>
/// <param name="AutosaveSlots">Server autosave slots, it is cycled through when the server autosaves.</param>
/// <param name="AfkAutokickInterval">How many minutes until someone is kicked when doing nothing.</param>
/// <param name="AutoPause">Whether should the server be paused when no players are present.</param>
/// <param name="OnlyAdminsCanPauseTheGame">Whether anyone or just admins can pause the game.</param>
/// <param name="AutosaveOnlyOnServer">Whether autosaves should be saved only on server or also on all connected clients.</param>
/// <param name="NonBlockingSaving">Highly experimental feature, enable only at your own risk of losing your saves. True for the server to fork itself to create an autosave. Autosaving on connected Windows clients will be disabled regardless of <paramref name="AutosaveOnlyOnServer"/> option.</param>
/// <param name="MinimumSegmentSize">Long network messages are split into segments that are sent over multiple ticks. Their size depends on the number of peers currently connected. Increasing the segment size will increase upload bandwidth requirement for the server and download bandwidth requirement for clients. This setting only affects server outbound messages. Changing these settings can have a negative impact on connection stability for some clients.</param>
/// <param name="MinimumSegmentSizePeerCount">Long network messages are split into segments that are sent over multiple ticks. Their size depends on the number of peers currently connected. Increasing the segment size will increase upload bandwidth requirement for the server and download bandwidth requirement for clients. This setting only affects server outbound messages. Changing these settings can have a negative impact on connection stability for some clients.</param>
/// <param name="MaximumSegmentSize">Long network messages are split into segments that are sent over multiple ticks. Their size depends on the number of peers currently connected. Increasing the segment size will increase upload bandwidth requirement for the server and download bandwidth requirement for clients. This setting only affects server outbound messages. Changing these settings can have a negative impact on connection stability for some clients.</param>
/// <param name="MaximumSegmentSizePeerCount">Long network messages are split into segments that are sent over multiple ticks. Their size depends on the number of peers currently connected. Increasing the segment size will increase upload bandwidth requirement for the server and download bandwidth requirement for clients. This setting only affects server outbound messages. Changing these settings can have a negative impact on connection stability for some clients.</param>
/// <returns></returns>
public record ServerSettings(
    string Name = "factoriod",
    string Description = "factoriod",
    IReadOnlyCollection<string>? Tags = null,
    uint? MaxPlayers = null,
    Visibility? Visibility = null,
    string Username = "",
    bool RequireUserVerification = true,
    uint? MaxUploadInKilobytesPerSecond = null,
    uint? MaxUploadSlots = null,
    uint? MinimumLatencyInTicks = null,

    [Range(6, 240)]
    uint MaxHeartbeatsPerSecond = 60,
    bool IgnorePlayerLimitForReturningPlayers = false,

    [RegularExpression(@"^((true)|(false)|(admins-only))", ErrorMessage = "Possible values are true, false, and admins-only")]
    string AllowCommands = "admins-only",
    uint AutosaveInterval = 10,
    uint AutosaveSlots = 5,
    uint? AfkAutokickInterval = null,
    bool AutoPause = true,
    bool OnlyAdminsCanPauseTheGame = true,
    bool AutosaveOnlyOnServer = true,
    bool NonBlockingSaving = false,
    uint MinimumSegmentSize = 25,
    uint MinimumSegmentSizePeerCount = 20,
    uint MaximumSegmentSize = 100,
    uint MaximumSegmentSizePeerCount = 10
)
{
    public IReadOnlyCollection<string> Tags { get; } = new PrintableReadOnlyCollection<string>(Tags ?? Array.Empty<string>());
    public uint? MaxPlayers { get; } = MaxPlayers ?? 0;
    public Visibility Visibility { get; } = Visibility ?? new();
    public uint? MaxUploadInKilobytesPerSecond { get; } = MaxUploadInKilobytesPerSecond ?? 0;
    public uint? MaxUploadSlots { get; } = MaxUploadSlots ?? 5;
    public uint? MinimumLatencyInTicks { get; } = MinimumLatencyInTicks ?? 0;
    public uint? AfkAutokickInterval { get; } = AfkAutokickInterval ?? 0;
}

/// <summary>
/// Factorio server settings, including secrets.
/// </summary>
/// <param name="Password">Your factorio.com login credentials. Required for games with <see cref="Visibility.Public"/></param>
/// <param name="Token">Authentication token. May be used instead of <paramref name="Password"/> above.</param>
/// <param name="GamePassword">The password to join the game.</param>
public record ServerSettingsWithSecrets(
    string Password = "",
    string Token = "",
    string GamePassword = ""
) : ServerSettings;

/// <summary>
/// Visibility of a game.
/// </summary>
/// <param name="Public">Game will be published on the official Factorio matching server</param>
/// <param name="Lan">Game will be broadcast on LAN</param>
public record Visibility(
    bool Public = true,
    bool Lan = true
);

File: crates/private/factorio-config/src/server_settings.rs
//! Server settings for a Factorio server.

use serde::{Deserialize, Serialize};

/// The server settings. These are reflected in the `server-settings.json` file and can be modified after generating a
/// map. The changes will take effect after the server is restarted.
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "snake_case")]
pub struct ServerSettings {
    /// Name of the game as it will appear in the game listing.
    pub name: String,

    /// Description of the game that will appear in the game listing.
    pub description: String,

    /// Tags that will be used to filter the game listing.
    pub tags: Vec<String>,

    /// Maximum number of players allowed, admins can join even a full server. 0 means unlimited.
    pub max_players: u32,

    /// Visibility of the game in the game listing.
    pub visibility: Visibility,

    /// Your factorio.com login username. Required for games with [`Visibility::public`] set to [`true`].
    pub username: String,

    /// When set to [`true`], the server will only allow clients that have a valid Factorio.com account.
    pub require_user_verification: bool,

    /// Optional. Default value is 0. 0 means unlimited.
    pub max_upload_in_kilobytes_per_second: u32,

    /// Optional. Default value is 5. 0 means unlimited.
    pub max_upload_slots: u32,

    /// Optional. One tick is 16ms in default speed. Default value is 0. 0 means no minimum.
    pub minimum_latency_in_ticks: u32,

    /// Network tick rate. Maximum rate game updates packets are sent at before bundling them together. Minimum value is 6, maximum value is 240.
    pub max_heartbeats_per_second: u32,

    /// Players that played on this map already can join even when the max player limit was reached.
    pub ignore_player_limit_for_returning_players: bool,

    /// Whether to allow players to use commands.
    pub allow_commands: AllowCommands,

    /// Autosave interval in minutes.
    pub autosave_interval: u32,

    /// Server autosave slots. It is cycled through when the server autosaves.
    pub autosave_slots: u32,

    /// How many minutes until someone is kicked when doing nothing. 0 is never.
    pub afk_autokick_interval: u32,

    /// Whether the server should be paused when no players are present.
    pub auto_pause: bool,

    pub only_admins_can_pause_the_game: bool,

    /// Whether autosaves should be saved only on server or also on all connected clients. Default is true.
    pub autosave_only_on_server: bool,

    /// Highly experimental feature, enable on at your own risk of losing your saves. On UNIX systems, server will fork
    /// itself to create an autosave. Autosaving on connected Windows clients will be disabled regardless of
    /// [`Self::autosave_only_on_server`] option.
    pub non_blocking_saving: bool,

    /// Long network messages are split into segments that are sent over multiple ticks. Their size depends on the
    /// number of peers currently connected. Increasing the segment size will increase upload bandwidth requirement for
    /// the server and download bandwidth requirement for clients. This setting only affects server outbound messages.
    /// Changing these settings can have a negative impact on connection stability for some clients.
    pub minimum_segment_size: u32,
    pub minimum_segment_size_peer_count: u32,
    pub maximum_segment_size: u32,
    pub maximum_segment_size_peer_count: u32,
}

impl Default for ServerSettings {
    fn default() -> Self {
        Self {
            name: "factoriod".to_owned(),
            description: "factoriod".to_owned(),
            tags: Vec::new(),
            max_players: 0,
            visibility: Visibility::default(),
            username: "".to_owned(),
            require_user_verification: true,
            max_upload_in_kilobytes_per_second: 0,
            max_upload_slots: 5,
            minimum_latency_in_ticks: 0,
            max_heartbeats_per_second: 60,
            ignore_player_limit_for_returning_players: false,
            allow_commands: AllowCommands::AdminsOnly,
            autosave_interval: 10,
            autosave_slots: 5,
            afk_autokick_interval: 0,
            auto_pause: true,
            only_admins_can_pause_the_game: true,
            autosave_only_on_server: true,
            non_blocking_saving: false,
            minimum_segment_size: 25,
            minimum_segment_size_peer_count: 20,
            maximum_segment_size: 100,
            maximum_segment_size_peer_count: 10,
        }
    }
}

pub struct ServerSettingsWithSecrets {
    pub server_settings: ServerSettings,

    /// Your factorio.com login credentials. Required for games with [`Visibility::public`] set to [`true`].
    pub password: String,

    /// Authentication token. May be used instead of [`Self::password`] for games with [`Visibility::public`] set to
    /// [`true`].
    pub token: String,

    /// Password for joining the game. Empty string means no password.
    pub game_password: String,
}

impl Default for ServerSettingsWithSecrets {
    fn default() -> Self {
        Self {
            server_settings: ServerSettings::default(),
            password: "".to_owned(),
            token: "".to_owned(),
            game_password: "".to_owned(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "snake_case")]
/// What commands are allowed in the game.
pub enum AllowCommands {
    /// All players can use commands.
    True,

    /// No commands are allowed.
    False,

    /// Only admins can use commands.
    AdminsOnly,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Visibility {
    /// Game will be published on the official Factorio matching server.
    pub public: bool,

    /// Game will be broadcast on LAN.
    pub lan: bool,
}

impl Default for Visibility {
    fn default() -> Self {
        Self {
            public: true,
            lan: true,
        }
    }
}
